src/io/image_io.cc

```diff
@@ -218,29 +218,29 @@ void Imread(const nnvm::NodeAttrs& attrs,
   const auto& param = nnvm::get<ImreadParam>(attrs.parsed);
 
   std::ifstream file(param.filename, std::ios::binary | std::ios::ate);
+  // if file is not open we get bad alloc after tellg
+  CHECK(file.is_open()) << "Imread: Couldn't open file: " << param.filename;
   size_t fsize = file.tellg();
   file.seekg(0, std::ios::beg);
-  auto buff = new uint8_t[fsize];
-  file.read(reinterpret_cast<char*>(buff), fsize);
+  auto buff = std::make_shared<std::vector<uint8_t> >(fsize);
```

Comment:
Using vector here unnecessary and is undesirable for a couple of reasons:
1) You don't need any high-level vector functionality.
2) It increases the compile time because templates are expensive to compile.
3) For some (or maybe all) implementations, the constructor will initialize all values, which are subsequently overwritten

Reply:
@cjolivier01 I think your points are right in theory but wrong in practice and on top if you consider the context of this function which is reading from disk. So I think you are missing the big picture.

The performance claims that you are making should not in 99% of the cases be a justification for writing unsafe code. This is obviously not a hot path of the code, is a seldom used function which does IO, which is orders of magnitude slower than the points you mentioned.

I think we can both agree that the small overhead that you mention doesn't matter at all in this case and it's better to have a buffer that will be deallocated and managed using RAII.

In any case, which is the alternative that you suggest?

Reply:
unsafe code? i’m not sure what you’re referring to. make_shared or uniqe_ptr will cause the buffer to be deleted. what are you claiming is “unsafe”.
by the way, performance is a big deal.

Reply:
Could you please explicitly type out how do you propose to allocate this buffer?

Reply:
std::unique_ptr<uint8_t[]> buff(new uint8_t[fsize]);

Reply:
@cjolivier01 thank you. As said previously, I believe it needs to be a shared_ptr since the pointer is passed to the engine and deallocated in a different thread. The array specialization of shared_ptr is not available until C++20, unlike your proposed solution with unique_ptr. Hence I suggest we accept the solution with a vector as a compromise right now. Agreed there's a small penalization on initializing the buffer, but this should not be a big deal given that we read from disk. Would you be willing to agree to this?

Reply:
you can put the unique ptr inside the shared ptr if you like, but putting a vector there burns CPU needlessly and steals it from other threads.  It's also overkill for a  simple buffer like this.

Reply:
You can also do something like:
std::shared_ptr<char> ptr(new char[256], std::default_delete<char[]>());

Reply:
shared_ptr can be replaced with a move capture. but I think its only available since c++14.